# Exercise 4.4: Data Wrangling & Subsetting

## **Section 1:** 
#### *Tasks within the lesson*

##### Information about the data descriptors used in the file:
###### df_ords = refers to the orders data 
###### df_prods = refers to the products data

In [265]:
# Importing packages
import pandas as pd
import numpy as np
import os

In [266]:
# Reading the data file:
path = r'/Users/balachandark/Desktop/Instacart Basket Analysis/02 Data'
df_ords = pd.read_csv(os.path.join(path, 'Original Data', 'orders.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col = False)
df_dep = pd.read_csv(os.path.join(path, 'Original Data', 'departments.csv'), index_col = False)


##### **Dropping columns and checking missing values:**

In [268]:
# Dropping the eval_set column:

df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [269]:
# Spotting missing values of columnb days_since_prior_order

df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

##### **Renaming columns:**

In [271]:
# Renaming column names
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [272]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

##### **Changing data type of variables:**

In [274]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [275]:
## Changing data type of column order_id

df_ords['order_id'] = df_ords['order_id'].astype('str')

##### **Transposing data**

In [277]:
## Transposing department data
df_dep_t = df_dep.T

In [278]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

##### **Resetting index**

In [280]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

##### **Setting new header**

In [282]:
# Accessing new header
new_header = df_dep_t.iloc[0]

# New department table without the first row
df_dep_t_new = df_dep_t[1:]

# Setting the header - departments table

df_dep_t_new.columns = new_header
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

##### **Creating dictionary from dataframe**

In [284]:
# Creating data dictionary
data_dict = df_dep_t_new.to_dict('index')
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [285]:
print(data_dict.get('19'))

{'department': 'snacks'}


##### **Subsetting dataframes**

In [287]:
# Subsetting products dataframe to find the snacks department

df_snacks =  df_prods[df_prods['department_id']==19]
df_snacks.head()

# Method two: using loc function

df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

# Method three:

df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

path

'/Users/balachandark/Desktop/Instacart Basket Analysis/02 Data'

##### **Exporting dataframes**

In [289]:
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'))

## **Section 2: Tasks**

#### **Question 2:** 
##### *Find another identifier variable in the df_ords dataframe* 

In [292]:
# Exploring the df_ords

df_ords.describe()

user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01        2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

#### **Question 3**
##### *Rename a column*

In [294]:
# Changing user_id as string data
df_ords['user_id'] = df_ords['user_id'].astype('str')

# Checking the data type changes
df_ords.dtypes

order_id                   object
user_id                    object
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [295]:
## Renaming the eval_set column to evaluation_set

df_ords.rename(columns = {'eval_set' : 'evaluation_set'}, inplace = True)

#### **Question 4**
##### *Busiest hours for placing orders*

In [297]:
## Finding the mode of the column order_hour_of_day

df_ords['order_hour_of_day'].value_counts()

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

##### **The most busiest hour of the day is 10 am.** 

#### **Question 5**
##### *What is the meaning behind a value of 4 in the 'department_id' column within df_prods?*

In [300]:
## Finding the meaning of value 4

print(data_dict.get('4'))

# df_prods[df_prods['department_id'] == 4]

{'department': 'produce'}


##### **The value 4 in the products dataframe represents 'produce'.** 

#### **Question 6**
##### *Create a subset of data with breakfast items*

In [303]:
## Subsetting breakfast items

prods_breakfast = df_prods[df_prods['department_id'] == 14]
prods_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

#### **Question 7**
##### *Create a subset of with all throw dinner parties items*

In [344]:
## Subsetting throw party items

#data_dict

df_throw_party = df_prods[df_prods['department_id'].isin([5 , 20, 7, 12])]
df_throw_party.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

#### **Question 8**
##### *How many rows does the last dataframe you created have?*

In [352]:
## Checking the number of rows of df_prods

df_prods.shape

(49693, 5)

##### **There are 49693 rows in the products dataframe.** 

#### **Question 9**
##### *user_id 1 - a strange customer?*

In [371]:
user_id_1 = df_ords[df_ords['user_id'] == '1']
user_id_1

order_id user_id evaluation_set  order_number  orders_day_of_week  \
0   2539329       1          prior             1                   2   
1   2398795       1          prior             2                   3   
2    473747       1          prior             3                   3   
3   2254736       1          prior             4                   4   
4    431534       1          prior             5                   4   
5   3367565       1          prior             6                   2   
6    550135       1          prior             7                   1   
7   3108588       1          prior             8                   1   
8   2295261       1          prior             9                   1   
9   2550362       1          prior            10                   4   
10  1187899       1          train            11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

##### **User ID 1 has a shopping pattern for ordering products on the Instacart. The customer orders early in the morning or in the**
##### **evenings. Further, predominantly orders once month and orders in the middle of week** 

#### **Question 10**
##### *user_id 1 - stats about the user*

In [378]:
df_ords[df_ords['user_id'] == '1'].describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

##### **There are 11 orders in the name of the user 1. And the orders were placed close to wednesdays in a week**
##### **There is an average of 19 days between each orders** 

#### **Question 12**
##### *Export the wrangled orders and departments dataframe to prepared data folder*

In [391]:
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'))
df_dep_t_new.to_csv(os.path.join(path, 'Prepared Data', 'departments_wrangled.csv'))

<center><b> End of the report </b></center>